src/cache.rs

```diff
@@ -27,9 +27,21 @@ impl LineCache {
         self.invalid_after
     }
 
+    /// Retrieve the line with the given number from the cache.
     pub fn get_line(&self, n: u64) -> Option<&Line> {
-        self.lines.get(&n)
+        // TODO this is a naive search by value. May want to make use of the
+        // contiguity of the cache lines (HashMap::get(<min of keys...> + n))
+        // for optimization, but need to establish cache contiguity checks
+        // first.
+        self.lines.iter().find_map(|(_,line)| {
+            if Some(n+1) == line.line_num {
+                Some(line)
+            } else {
+                None
+            }
+        })
```

Comment:
Should frontends really try to use the actual line number here? At least tau just uses the linenumber (with linebreaks), since it's pretty hard and expensive to always convert back and forth from actual linenumbers to the linennumber with linebreaks

Reply:
Please completely ignore this change. I was making a wrong assumption and I reverted it shortly after I pushed it to my branch.

Looks like there was another problem in the function get_line(), which is a new one, and which wasn't used in xi-term (that's why I didn't see any crash earlier). I've just pushed another update for this fn's logic (it needs to take the invalid lines into consideration). Unfortunately that doesn't fix the cache corruption in tau.

Re. this cache corruption: I've added a new assert in xrl, which basically checks whether the cache line numbers form a contiguous sequence. Tau currently fails this assertion if you make a few scrolls and then click in the middle of the window. Looking at the cache state, I can say that this is triggered by a very strange update from xi-core:

copy X lines
invalidate 1 line
skip 1 line
copy Y lines

Xi-core is basically requesting LineCache to make a hole (1 invalidated line) in the middle of the cache. Do you know whether this is legal? LineCache is definitely not prepared to handle such situation. This is unique for tau, I could not make this update show up in xi-term.

Reply:
Hm, weird that it only happens in Tau, not quite sure where it'd differ from xi-term :thinking: 

I think it'd be best to ask in Xi's Zulip about this